apisix/balancer/ewma.lua

```diff
@@ -35,73 +58,69 @@ local function decay_ewma(ewma, last_touched_at, rtt, now)
     return ewma
 end
 
-
 local function store_stats(upstream, ewma, now)
     local success, err, forcible = shm_last_touched_at:set(upstream, now)
     if not success then
-        core.log.error("balancer_ewma_last_touched_at:set failed ", err)
+        core.log.error("shm_last_touched_at:set failed: ", err)
     end
     if forcible then
-        core.log.warn("balancer_ewma_last_touched_at:set valid items forcibly overwritten")
+        core.log.warn("shm_last_touched_at:set valid items forcibly overwritten")
     end
 
     success, err, forcible = shm_ewma:set(upstream, ewma)
     if not success then
-        core.log.error("balancer_ewma:set failed ", err)
+        core.log.error("shm_ewma:set failed: ", err)
     end
     if forcible then
-        core.log.warn("balancer_ewma:set valid items forcibly overwritten")
+        core.log.warn("shm_ewma:set valid items forcibly overwritten")
     end
 end
 
-
 local function get_or_update_ewma(upstream, rtt, update)
+    local lock_err = nil
+    if update then
+        lock_err = lock(upstream)
+    end
+
     local ewma = shm_ewma:get(upstream) or 0
+    if lock_err ~= nil then
```

Comment:
Why check `lock_err` after fetching `ewma`? It's so strange we don't check the return value after `lock` return. Or you should get `ewma` before fetching the lock.

Reply:
We need lock when update, and do not lock when get.

Reply:
We need to return ewma value for get operation.